tendermint/src/block/commit_sig.rs

```diff
@@ -1,78 +1,182 @@
+                                Err(A::Error::custom(format!(
+                                    "signature is present for BlockIDFlagAbsent CommitSig {:?}",
+                                    incoming.signature.unwrap()
+                                )))
+                            } else {
+                                Ok(CommitSig::BlockIDFlagAbsent {
+                                    validator_address: incoming.validator_address,
+                                })
+                            }
+                        }
+                    }
+                    // BlockIDFlagCommit
+                    2 => {
+                        if incoming.timestamp.is_none() {
+                            Err(A::Error::custom(
+                                "timestamp is null for BlockIDFlagCommit CommitSig",
+                            ))
+                        } else {
+                            if incoming.signature.is_none() {
+                                Err(A::Error::custom(
+                                    "signature is null for BlockIDFlagCommit CommitSig",
+                                ))
+                            } else {
+                                Ok(CommitSig::BlockIDFlagCommit {
+                                    validator_address: incoming.validator_address,
+                                    timestamp: incoming.timestamp.unwrap(),
+                                    signature: incoming.signature.unwrap(),
+                                })
+                            }
+                        }
+                    }
+                    // BlockIDFlagNil
+                    3 => {
+                        if incoming.timestamp.is_none() {
+                            Err(A::Error::custom(
+                                "timestamp is null for BlockIDFlagNil CommitSig",
+                            ))
+                        } else {
+                            if incoming.signature.is_none() {
+                                Err(A::Error::custom(
+                                    "signature is null for BlockIDFlagNil CommitSig",
+                                ))
+                            } else {
+                                Ok(CommitSig::BlockIDFlagNil {
+                                    validator_address: incoming.validator_address,
+                                    timestamp: incoming.timestamp.unwrap(),
+                                    signature: incoming.signature.unwrap(),
+                                })
+                            }
+                        }
+                    }
+                    // Error: unknown CommitSig type
+                    e => Err(A::Error::custom(format!(
+                        "unknown BlockIdFlag value in CommitSig {}",
+                        e
+                    ))),
+                }
+            }
+        }
+
+        deserializer.deserialize_map(CommitSigVisitor)
```

Comment:
While the enum idea is cool, it seems to come with the cost of a rather complicated `Deserialize` impl for `commitSig`.

Reply:
Yes and that's the tradeoff @brapse was going to go for in an earlier discussion. (Feel free to engage Sean.) The idea is that serialization is not actually the core product here and during the usual development it's just annoying to do something some particular way just because serialization says so.
This approach puts the logic of making sure that the incoming data is deserialized properly (with some basic validation) and the outgoing data is serialized properly in the serialization "layer" and the internal workings of the application can become completely independent from it. For example it can make use of the additional features of Rust better.
Just to highlight it even more: if/when the Tendermint data model changes because the Go + Rust teams decide on some other implementation, the changes can be implemented in this layer without affecting the application.

Reply:
I have to admit there is _one_ not-completely-Rust-native-best-practice feature in the code and it annoys me slightly. I deserialize the incoming CommitSig into an internal CommitSigSpec structure that is defined in ADR-025. I could have gone hard-core and implement complete deserialization while parsing the incoming text. But that would be even more convoluted and harder to change down the road. I feel this is a good compromise that allows us to easily update the deserialization code, if a new ADR comes by.

Reply:
> Yes and that's the tradeoff @brapse was going to go for in an earlier discussion. (Feel free to engage Sean.) The idea is that serialization is not actually the core product here and during the usual development it's just annoying to do something some particular way just because serialization says so.

That is absolutely reasonable. Serialization isn't the core product and we shouldn't build our types around it. All in favour of decoupling serialization concerns and core types.

What I was referring to was that the seemingly beautiful change made the part where the serialization actually happens difficult to read and quite large. A simpler way might be to have a serialization type and a `From` / `Into` for that and the actual type. This is very similar to what you are already doing with the `CommitSigSpec` type but simpler to parse for a human :-)

Reply:
Done and done: Today's commit moved things into serde's "from" annotation. So now, the code looks a bit cleaner, although it's not less code:
In the serializers, a new RawCommitSig type was introduced that deserializes from JSON and it looks very much like the definition in the ADR (including the enum for BlockIDFlags). Deserialization automatically invokes basic type-checking of the struct (like an u8 has to be an u8).
The Rust-native CommitSig type converts from this RawCommitSig type through it's "From" trait which holds the validation for the struct. This logic is more about "if this variable is set, than that other variable also has to have a value".

So the deserializer does type-check and the From trait does validation. My only grief is that type-checking happens in the serializers library while validation happens in the struct's library (in this case at CommitSig). Interestingly, this is similar to how it used to be, so I'm guessing the developers will be fine with it.

Reply:
If the code looks cleaner that is already worth it 👍  And I think it is OK for the core type to defines when it's valid (s.t. you can only construct a valid type).